File: csharp/HearthstoneReplays/Parser/Handlers/DataHandler.cs
#region

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using HearthDb.Enums;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.Entities;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using HearthstoneReplays.Parser.ReplayData.Meta;
using Action = HearthstoneReplays.Parser.ReplayData.GameActions.Action;

#endregion

namespace HearthstoneReplays.Parser.Handlers
{
	public class DataHandler
	{
		public static void Handle(string timestamp, string data, ParserState state)
		{
			var trimmed = data.Trim();
			var indentLevel = data.Length - trimmed.Length;
			data = trimmed;

			if(state.Node != null && indentLevel <= state.Node.IndentLevel)
			{
				var action = state.Node.Object as Action;
				if(action == null || action.Entity != 1 || !data.ToLower().Contains("mulligan"))
					state.Node = state.Node.Parent ?? state.Node;
			}


			if(data == "CREATE_GAME")
			{
				state.CurrentGame = new Game {Data = new List<GameData>()};
				state.Replay.Games.Add(state.CurrentGame);
				state.Node = new Node(typeof(Game), state.CurrentGame, 0, null);
				return;
			}

			var match = Regexes.ActionCreategameRegex.Match(data);
			if(match.Success)
			{
				var id = match.Groups[1].Value;
				Debug.Assert(id == "1");
				var gEntity = new GameEntity {Id = int.Parse(id), Tags = new List<Tag>()};
				state.CurrentGame.Data.Add(gEntity);
				state.Node = new Node(typeof(GameEntity), gEntity, indentLevel, state.Node);
				return;
			}

			match = Regexes.ActionCreategamePlayerRegex.Match(data);
			if(match.Success)
			{
				var id = match.Groups[1].Value;
				var playerId = match.Groups[2].Value;
				var accountHi = match.Groups[3].Value;
				var accountLo = match.Groups[4].Value;
				var pEntity = new PlayerEntity
				{
					Id = int.Parse(id),
					AccountHi = accountHi,
					AccountLo = accountLo,
					PlayerId = int.Parse(playerId),
					Tags = new List<Tag>()
				};
				state.CurrentGame.Data.Add(pEntity);
				state.Node = new Node(typeof(PlayerEntity), pEntity, indentLevel, state.Node);
				return;
			}

			match = Regexes.ActionStartRegex.Match(data);
			if(match.Success)
			{
				var rawEntity = match.Groups[1].Value;
				var rawType = match.Groups[2].Value;
				var index = match.Groups[3].Value;
				var rawTarget = match.Groups[4].Value;
				var entity = Helper.ParseEntity(rawEntity, state);
				var target = Helper.ParseEntity(rawTarget, state);
				var type = Helper.ParseEnum<PowSubType>(rawType);
				var action = new Action
				{
					Data = new List<GameData>(),
					Entity = entity,
					Index = int.Parse(index),
					Target = target,
					//TimeStamp = timestamp,
					Type = type
				};
				if(state.Node.Type == typeof(Game))
					((Game)state.Node.Object).Data.Add(action);
				else if(state.Node.Type == typeof(Action))
					((Action)state.Node.Object).Data.Add(action);
				else
					throw new Exception("Invalid node " + state.Node.Type);
				state.Node = new Node(typeof(Action), action, indentLevel, state.Node);
				return;
			}

			match = Regexes.ActionMetadataRegex.Match(data);
			if(match.Success)
			{
				var rawMeta = match.Groups[1].Value;
				var rawData = match.Groups[2].Value;
				var info = match.Groups[3].Value;
				var parsedData = Helper.ParseEntity(rawData, state);
				var meta = Helper.ParseEnum<MetaDataType>(rawMeta);
				var metaData = new MetaData {Data = parsedData, Info = int.Parse(info), Meta = meta, MetaInfo = new List<Info>()};
				if(state.Node.Type == typeof(Action))
					((Action)state.Node.Object).Data.Add(metaData);
				else
					throw new Exception("Invalid node " + state.Node.Type);
				state.Node = new Node(typeof(MetaData), metaData, indentLevel, state.Node);
				return;
			}

			match = Regexes.ActionMetaDataInfoRegex.Match(data);
			if(match.Success)
			{
				var index = match.Groups[1].Value;
				var rawEntity = match.Groups[2].Value;
				var entity = Helper.ParseEntity(rawEntity, state);
				var metaInfo = new Info {Id = entity, Index = int.Parse(index)};
				if(state.Node.Type == typeof(MetaData))
					((MetaData)state.Node.Object).MetaInfo.Add(metaInfo);
				else
					throw new Exception("Invalid node " + state.Node.Type);
			}

			match = Regexes.ActionShowEntityRegex.Match(data);
			if(match.Success)
			{
				var rawEntity = match.Groups[1].Value;
				var cardId = match.Groups[2].Value;
				var entity = Helper.ParseEntity(rawEntity, state);
				var showEntity = new ShowEntity {CardId = cardId, Entity = entity, Tags = new List<Tag>()};
				if(state.Node.Type == typeof(Game))
					((Game)state.Node.Object).Data.Add(showEntity);
				else if(state.Node.Type == typeof(Action))
					((Action)state.Node.Object).Data.Add(showEntity);
				else
					throw new Exception("Invalid node " + state.Node.Type);
				state.Node = new Node(typeof(ShowEntity), showEntity, indentLevel, state.Node);
				return;
			}

			match = Regexes.ActionHideEntityRegex.Match(data);
			if(match.Success)
			{
				var rawEntity = match.Groups[1].Value;
				var tagName = match.Groups[2].Value;
				var value = match.Groups[3].Value;
				var entity = Helper.ParseEntity(rawEntity, state);
				var zone = Helper.ParseTag(tagName, value);
				var hideEntity = new HideEntity {Entity = entity, Zone = zone.Value};
				if(state.Node.Type == typeof(Game))
					((Game)state.Node.Object).Data.Add(hideEntity);
				else if(state.Node.Type == typeof(Action))
					((Action)state.Node.Object).Data.Add(hideEntity);
				else
					throw new Exception("Invalid node: " + state.Node.Type);
				return;
			}

			match = Regexes.ActionFullEntityUpdatingRegex.Match(data);
			if(!match.Success)
				match = Regexes.ActionFullEntityCreatingRegex.Match(data);
			if(match.Success)
			{
				var rawEntity = match.Groups[1].Value;
				var cardId = match.Groups[2].Value;
				var entity = Helper.ParseEntity(rawEntity, state);
				var showEntity = new FullEntity {CardId = cardId, Id = entity, Tags = new List<Tag>()};
				if(state.Node.Type == typeof(Game))
					((Game)state.Node.Object).Data.Add(showEntity);
				else if(state.Node.Type == typeof(Action))
					((Action)state.Node.Object).Data.Add(showEntity);
				else
					throw new Exception("Invalid node " + state.Node.Type);
				state.Node = new Node(typeof(FullEntity), showEntity, indentLevel, state.Node);
				return;
			}

			match = Regexes.ActionTagChangeRegex.Match(data);
			if(match.Success)
			{
				var rawEntity = match.Groups[1].Value;
				var tagName = match.Groups[2].Value;
				var value = match.Groups[3].Value;
				var tag = Helper.ParseTag(tagName, value);
                if(tag.Name == (int)GameTag.CURRENT_PLAYER)
                    UpdateCurrentPlayer(state, rawEntity, tag);
                var entity = Helper.ParseEntity(rawEntity, state);
                if(tag.Name == (int)GameTag.ENTITY_ID)
                    entity = UpdatePlayerEntity(state, rawEntity, tag, entity);
                var tagChange = new TagChange {Entity = entity, Name = tag.Name, Value = tag.Value};
				if(state.Node.Type == typeof(Game))
					((Game)state.Node.Object).Data.Add(tagChange);
				else if(state.Node.Type == typeof(Action))
					((Action)state.Node.Object).Data.Add(tagChange);
				else
					throw new Exception("Invalid node " + state.Node.Type);
				return;
			}

			match = Regexes.ActionTagRegex.Match(data);
			if(match.Success)
			{
				var tagName = match.Groups[1].Value;
				var value = match.Groups[2].Value;
				var tag = Helper.ParseTag(tagName, value);
				if(tag.Name == (int)GameTag.CURRENT_PLAYER)
					state.FirstPlayerId = ((PlayerEntity)state.Node.Object).Id;
				if(state.Node.Type == typeof(GameEntity))
					((GameEntity)state.Node.Object).Tags.Add(tag);
				else if(state.Node.Type == typeof(PlayerEntity))
					((PlayerEntity)state.Node.Object).Tags.Add(tag);
				else if(state.Node.Type == typeof(FullEntity))
					((FullEntity)state.Node.Object).Tags.Add(tag);
				else if(state.Node.Type == typeof(ShowEntity))
					((ShowEntity)state.Node.Object).Tags.Add(tag);
				else
					throw new Exception("Invalid node " + state.Node.Type + " -- " + data);
			}
		}

        private static int UpdatePlayerEntity(ParserState state, string rawEntity, Tag tag, int entity)
        {
            int tmp;
            if(!int.TryParse(rawEntity, out tmp) && !rawEntity.StartsWith("[") && rawEntity != "GameEntity")
            {
                if(entity != tag.Value)
                {
                    entity = tag.Value;
                    var tmpName = ((PlayerEntity)state.CurrentGame.Data[1]).Name;
                    ((PlayerEntity)state.CurrentGame.Data[1]).Name =
                        ((PlayerEntity)state.CurrentGame.Data[2]).Name;
                    ((PlayerEntity)state.CurrentGame.Data[2]).Name = tmpName;
                    foreach(var dataObj in ((Game)state.Node.Object).Data)
                    {
                        var tChange = dataObj as TagChange;
                        if(tChange != null)
                            tChange.Entity = tChange.Entity == 2 ? 3 : 2;
                    }
                }
            }

            return entity;
        }

        private static void UpdateCurrentPlayer(ParserState state, string rawEntity, Tag tag)
        {
            if(tag.Value == 0)
            {
                try
                {
                    Helper.ParseEntity(rawEntity, state);
                }
                catch
                {
                    var currentPlayer =
                        (PlayerEntity)state.CurrentGame.Data.Single(x => (x is PlayerEntity) && ((PlayerEntity)x).Id == state.CurrentPlayerId);
                    currentPlayer.Name = rawEntity;
                }
            }
            else if(tag.Value == 1)
            {
                try
                {
                    Helper.ParseEntity(rawEntity, state);
                }
                catch
                {
                    var currentPlayer =
                        (PlayerEntity)state.CurrentGame.Data.Single(x => (x is PlayerEntity) && ((PlayerEntity)x).Id != state.CurrentPlayerId);
                    currentPlayer.Name = rawEntity;
                }
                state.CurrentPlayerId = Helper.ParseEntity(rawEntity, state);
            }
        }
    }
}

File: java/src/main/java/info/hearthsim/hsreplay/parser/handlers/DataHandler.java
package info.hearthsim.hsreplay.parser.handlers;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;

import info.hearthsim.hsreplay.enums.GameTag;
import info.hearthsim.hsreplay.enums.MetaDataType;
import info.hearthsim.hsreplay.enums.PowSubType;
import info.hearthsim.hsreplay.parser.Helper;
import info.hearthsim.hsreplay.parser.Node;
import info.hearthsim.hsreplay.parser.ParserState;
import info.hearthsim.hsreplay.parser.Regexes;
import info.hearthsim.hsreplay.parser.Utils;
import info.hearthsim.hsreplay.parser.replaydata.Game;
import info.hearthsim.hsreplay.parser.replaydata.GameData;
import info.hearthsim.hsreplay.parser.replaydata.entities.FullEntity;
import info.hearthsim.hsreplay.parser.replaydata.entities.GameEntity;
import info.hearthsim.hsreplay.parser.replaydata.entities.PlayerEntity;
import info.hearthsim.hsreplay.parser.replaydata.gameactions.Action;
import info.hearthsim.hsreplay.parser.replaydata.gameactions.Action.ActionBuilder;
import info.hearthsim.hsreplay.parser.replaydata.gameactions.HideEntity;
import info.hearthsim.hsreplay.parser.replaydata.gameactions.ShowEntity;
import info.hearthsim.hsreplay.parser.replaydata.gameactions.Tag;
import info.hearthsim.hsreplay.parser.replaydata.gameactions.TagChange;
import info.hearthsim.hsreplay.parser.replaydata.meta.Info;
import info.hearthsim.hsreplay.parser.replaydata.meta.MetaData;
import lombok.extern.slf4j.Slf4j;

@Slf4j
public class DataHandler {

	private static int previousTimestampHours;

	public static void handle(String timestamp, String data, ParserState state) throws Exception {

		timestamp = normalizeTimestamp(timestamp);

		String trimmed = data.trim();
		int indentLevel = data.length() - trimmed.length();
		data = trimmed;

		log.info("Considering line " + data);

		if ("CREATE_GAME".equals(data)) {
			Game currentGame = Game.builder().data(new ArrayList<GameData>()).timestamp(timestamp).build();
			state.setCurrentGame(currentGame);
			state.getReplay().getGames().add(currentGame);
			Node node = new Node(Game.class, currentGame, 0, null);
			state.setNode(node);
			return;
		}

		// TODO: special case where the CREATE_GAME tag is not present in the
		// output log
		if (state.getReplay().getGames().isEmpty()) {
			Game currentGame = Game.builder().data(new ArrayList<GameData>()).timestamp(timestamp).build();
			state.setCurrentGame(currentGame);
			state.getReplay().getGames().add(currentGame);
			Node node = new Node(Game.class, currentGame, 0, null);
			state.setNode(node);
		}

		if ("ACTION_END".equals(data)) {
			state.updateCurrentNode(Game.class, Action.class);
			// log.info("Setting state node. Parent is " +
			// state.getNode().getParent());
			// log.info("Current node is " + state.getNode());
			state.setNode(state.getNode().getParent() != null ? state.getNode().getParent() : state.getNode());
			// log.info("New node is " + state.getNode());
			return;
		}

		Matcher match = Regexes.ActionCreategameRegex.matcher(data);
		if (match.matches()) {
			String id = match.group(1);
			if (Integer.parseInt(id) != 1) { throw new Exception("Game ID incorrect! " + id); }
			GameEntity gEntity = new GameEntity(Integer.parseInt(id), new ArrayList<Tag>());
			state.getCurrentGame().getData().add(gEntity);
			Node node = new Node(GameEntity.class, gEntity, indentLevel, state.getNode());
			state.setNode(node);
			return;
		}

		match = Regexes.ActionCreategamePlayerRegex.matcher(data);
		if (match.matches()) {
			String id = match.group(1);
			String playerId = match.group(2);
			String accountHi = match.group(3);
			String accountLo = match.group(4);
			PlayerEntity pEntity = PlayerEntity.builder().accountHi(accountHi).accountLo(accountLo)
					.playerId(Integer.parseInt(playerId)).build();
			pEntity.setTags(new ArrayList<Tag>());
			pEntity.setId(Integer.parseInt(id));
			state.updateCurrentNode(Game.class);
			state.getCurrentGame().getData().add(pEntity);
			Node node = new Node(PlayerEntity.class, pEntity, indentLevel, state.getNode());
			state.setNode(node);
			return;
		}

		match = Regexes.ActionStartRegex_OLD.matcher(data);
		if (match.matches()) {
			String rawEntity = match.group(1);
			String rawType = match.group(2);
			String index = match.group(3);
			String rawTarget = match.group(4);
			int entity = Helper.parseEntity(rawEntity, state);
			int target = Helper.parseEntity(rawTarget, state);
			int type = PowSubType.parseEnum(rawType);

			Action action = Action.builder().data(new ArrayList<GameData>()).entity(entity)
					.index(Integer.parseInt(index)).target(target).timestamp(timestamp).type(type).build();
			state.updateCurrentNode(Game.class, Action.class);

			if (state.getNode().getType().isAssignableFrom(Game.class)) {
				((Game) state.getNode().getObject()).getData().add(action);
			}
			else if (state.getNode().getType().isAssignableFrom(Action.class)) {
				((Action) state.getNode().getObject()).getData().add(action);
			}
			else {
				throw new Exception("Invalid node " + state.getNode().getType());
			}

			Node node = new Node(Action.class, action, indentLevel, state.getNode());
			state.setNode(node);
			return;
		}

		match = Regexes.ActionStartRegex.matcher(data);
		if (match.matches()) {
			String rawType = match.group(1);
			String rawEntity = match.group(2);
			String effectId = match.group(3);
			String effectIndex = match.group(4);
			String rawTarget = match.group(5);
			int entity = Helper.parseEntity(rawEntity, state);
			int target = Helper.parseEntity(rawTarget, state);
			int type = PowSubType.parseEnum(rawType);

			ActionBuilder builder = Action.builder().data(new ArrayList<GameData>()).entity(entity).target(target)
					.timestamp(timestamp).type(type);

			if (effectId != null && effectId.length() > 0) {
				builder.effectId(Integer.parseInt(effectId));
			}
			if (effectIndex != null && effectIndex.length() > 0) {
				builder.effectIndex(Integer.parseInt(effectIndex));
			}

			Action action = builder.build();
			state.updateCurrentNode(Game.class, Action.class);

			if (state.getNode().getType().isAssignableFrom(Game.class)) {
				((Game) state.getNode().getObject()).getData().add(action);
			}
			else if (state.getNode().getType().isAssignableFrom(Action.class)) {
				((Action) state.getNode().getObject()).getData().add(action);
			}
			else {
				throw new Exception("Invalid node " + state.getNode().getType());
			}

			Node node = new Node(Action.class, action, indentLevel, state.getNode());
			state.setNode(node);
			return;
		}

		match = Regexes.ActionMetadataRegex.matcher(data);
		if (match.matches()) {
			String rawMeta = match.group(1);
			String rawData = match.group(2);
			String info = match.group(3);
			int parsedData = Helper.parseEntity(rawData, state);
			int meta = MetaDataType.parseEnum(rawMeta);

			MetaData metaData = MetaData.builder().data(parsedData).info(Integer.parseInt(info)).meta(meta).entity(meta)
					.timestamp(timestamp).metaInfo(new ArrayList<Info>()).build();
			state.updateCurrentNode(Action.class);
			log.debug("\tHandling metaData " + metaData + " on node " + state.getNode().getType());

			if (state.getNode().getType().isAssignableFrom(Action.class)) {
				((Action) state.getNode().getObject()).getData().add(metaData);
			}
			else {
				throw new Exception("Invalid node " + state.getNode().getType() + " for data " + data);
			}

			Node node = new Node(MetaData.class, metaData, indentLevel, state.getNode());
			state.setNode(node);
			return;
		}

		match = Regexes.ActionMetaDataInfoRegex.matcher(data);
		if (match.matches()) {
			String index = match.group(1);
			String rawEntity = match.group(2);
			int entity = Helper.parseEntity(rawEntity, state);
			Info metaInfo = new Info(Integer.parseInt(index), entity, entity);

			if (state.getNode().getType().isAssignableFrom(MetaData.class)) {
				((MetaData) state.getNode().getObject()).getMetaInfo().add(metaInfo);
			}
			else {
				throw new Exception("Invalid node " + state.getNode().getType());
			}

			return;
		}

		match = Regexes.ActionShowEntityRegex.matcher(data);
		if (match.matches()) {
			String rawEntity = match.group(1);
			String cardId = match.group(2);
			int entity = Helper.parseEntity(rawEntity, state);

			ShowEntity showEntity = ShowEntity.builder().cardId(cardId).entity(entity).tags(new ArrayList<Tag>())
					.timestamp(timestamp).build();
			state.updateCurrentNode(Game.class, Action.class);

			if (state.getNode().getType().isAssignableFrom(Game.class)) {
				((Game) state.getNode().getObject()).getData().add(showEntity);
			}
			else if (state.getNode().getType().isAssignableFrom(Action.class)) {
				((Action) state.getNode().getObject()).getData().add(showEntity);
			}
			else {
				throw new Exception("Invalid node " + state.getNode().getType());
			}

			Node node = new Node(ShowEntity.class, showEntity, indentLevel, state.getNode());
			state.setNode(node);
			return;
		}

		match = Regexes.ActionHideEntityRegex.matcher(data);
		if (match.matches()) {
			String rawEntity = match.group(1);
			String tagName = match.group(2);
			String value = match.group(3);
			int entity = Helper.parseEntity(rawEntity, state);
			Tag zone = Helper.parseTag(tagName, value);

			HideEntity hideEntity = new HideEntity(timestamp, entity, zone.getValue());
			state.updateCurrentNode(Game.class, Action.class);

			if (state.getNode().getType().isAssignableFrom(Game.class)) {
				((Game) state.getNode().getObject()).getData().add(hideEntity);
			}
			else if (state.getNode().getType().isAssignableFrom(Action.class)) {
				((Action) state.getNode().getObject()).getData().add(hideEntity);
			}
			else {
				throw new Exception("Invalid node " + state.getNode().getType());
			}
			return;
		}

		match = Regexes.ActionFullEntityUpdatingRegex.matcher(data);
		if (!match.matches()) {
			match = Regexes.ActionFullEntityCreatingRegex.matcher(data);
		}

		if (match.matches()) {
			String rawEntity = match.group(1);
			String cardId = match.group(2);
			int entity = Helper.parseEntity(rawEntity, state);

			FullEntity fullEntity = new FullEntity(timestamp, cardId, entity, new ArrayList<Tag>());
			state.updateCurrentNode(Game.class, Action.class);

			if (state.getNode().getType().isAssignableFrom(Game.class)) {
				((Game) state.getNode().getObject()).getData().add(fullEntity);
			}
			else if (state.getNode().getType().isAssignableFrom(Action.class)) {
				((Action) state.getNode().getObject()).getData().add(fullEntity);
			}
			else {
				throw new Exception("Invalid node " + state.getNode().getType());
			}

			Node node = new Node(FullEntity.class, fullEntity, indentLevel, state.getNode());
			state.setNode(node);
			return;
		}

		match = Regexes.ActionTagChangeRegex.matcher(data);
		if (match.matches()) {
			String rawEntity = match.group(1);
			String tagName = match.group(2);
			String value = match.group(3);
			Tag tag = Helper.parseTag(tagName, value);

			if (tag.getName() == GameTag.CURRENT_PLAYER.getIntValue()) {
				updateCurrentPlayer(state, rawEntity, tag);
			}
			int entity = Helper.parseEntity(rawEntity, state);
			if (tag.getName() == GameTag.ENTITY_ID.getIntValue()) {
				entity = updatePlayerEntity(state, rawEntity, tag, entity);
			}

			TagChange tagChange = new TagChange(entity, tag.getName(), tag.getValue());
			state.updateCurrentNode(Game.class, Action.class);
			log.info("\thandling tag change " + tagChange + " on current node " + state.getNode().getType());

			if (state.getNode().getType().isAssignableFrom(Game.class)) {
				((Game) state.getNode().getObject()).getData().add(tagChange);
			}
			else if (state.getNode().getType().isAssignableFrom(Action.class)) {
				((Action) state.getNode().getObject()).getData().add(tagChange);
			}
			else {
				throw new Exception("Invalid node " + state.getNode().getType());
			}

			return;
		}

		match = Regexes.ActionTagRegex.matcher(data);
		if (match.matches()) {
			String tagName = match.group(1);
			String value = match.group(2);
			Tag tag = Helper.parseTag(tagName, value);

			if (tag.getName() == GameTag.CURRENT_PLAYER.getIntValue()) {
				state.setFirstPlayerId(((PlayerEntity) state.getNode().getObject()).getId());
			}

			if (state.getNode().getType().isAssignableFrom(GameEntity.class)) {
				((GameEntity) state.getNode().getObject()).getTags().add(tag);
			}
			else if (state.getNode().getType().isAssignableFrom(PlayerEntity.class)) {
				((PlayerEntity) state.getNode().getObject()).getTags().add(tag);
			}
			else if (state.getNode().getType().isAssignableFrom(FullEntity.class)) {
				((FullEntity) state.getNode().getObject()).getTags().add(tag);
			}
			else if (state.getNode().getType().isAssignableFrom(ShowEntity.class)) {
				((ShowEntity) state.getNode().getObject()).getTags().add(tag);
			}
			else {
				throw new Exception("Invalid node " + state.getNode().getType());
			}
		}
	}

	private static String normalizeTimestamp(String timestamp) {
		String[] split = timestamp.split(":");
		int hours = Integer.parseInt(split[0]);
		if (hours < previousTimestampHours) {
			log.debug("Computed new ts " + hours);
			hours = previousTimestampHours + 1;
			String newTs = hours + ":" + split[1] + ":" + split[2];
			return newTs;
		}
		previousTimestampHours = hours;
		return timestamp;
	}

	private static int updatePlayerEntity(ParserState state, String rawEntity, Tag tag, int entity) {
		if (!Utils.isInteger(rawEntity) && !rawEntity.startsWith("[") && !rawEntity.equals("GameEntity")) {
			if (entity != tag.getValue()) {
				entity = tag.getValue();
				String tmpName = ((PlayerEntity) state.getCurrentGame().getData().get(1)).getName();

				((PlayerEntity) state.getCurrentGame().getData().get(1))
						.setName(((PlayerEntity) state.getCurrentGame().getData().get(2)).getName());
				((PlayerEntity) state.getCurrentGame().getData().get(2)).setName(tmpName);

				for (Object obj : ((Game) state.getNode().getObject()).getData()) {
					TagChange tChange = obj instanceof TagChange ? (TagChange) obj : null;
					if (tChange != null) {
						tChange.setEntity(tChange.getEntity() == 2 ? 3 : 2);
					}
				}
			}
		}
		return entity;
	}

	private static void updateCurrentPlayer(ParserState state, String rawEntity, Tag tag) throws Exception {
		List<GameData> data = state.getCurrentGame().getData();
		if (tag.getValue() == 0) {
			try {
				Helper.parseEntity(rawEntity, state);
			}
			catch (Exception e) {
				for (GameData x : data) {
					if (x instanceof PlayerEntity && ((PlayerEntity) x).getId() == state.getCurrentPlayerId()) {
						((PlayerEntity) x).setName(rawEntity);
						break;
					}
				}
			}
		}
		else if (tag.getValue() == 1) {
			try {
				Helper.parseEntity(rawEntity, state);
			}
			catch (Exception e) {
				for (GameData x : data) {
					if (x instanceof PlayerEntity && ((PlayerEntity) x).getId() != state.getCurrentPlayerId()) {
						((PlayerEntity) x).setName(rawEntity);
						break;
					}
				}
			}
			state.setCurrentPlayerId(Helper.parseEntity(rawEntity, state));
		}
	}
}
